include/envoy/request_info/dynamic_metadata.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <atomic>
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace RequestInfo {
+
+class DynamicMetadata {
+public:
+  virtual ~DynamicMetadata(){};
+
+  /**
+   * @param data_name the name of the data being set.
+   * @param data an owning pointer to the data to be stored.
+   * Note that it is an error to call setData() twice with the same data_name.
```

Comment:
Is this limitation going to be an issue anywhere?

Reply:
I presume you mean in the currently existing uses of dynamic metadata?  

The simple answer is "I don't know--haven't really started on the conversion yet".  The more complex answer is:
* This was put in as a way to keep the use of dynamic metadata under control, specifically, to make sure without dependency management (since that implementation comes later) that only one "source" sets any particular piece of data.
* The structure that allows for lazy evaluation also allows for the dynamic metadata source to change DM entries without that needing to go through the DM interface.  So if there are single sources that want to change DM over the lifetime of the request, that can still be implemented within this restriction.  The only thing this really forbids is something setting a piece of data and then something else overriding it.

Reply:
As we discussed offline, I expanded the comment here to give more of an explanation for this choice.  I looked at STYLE.md to see if adding something there made sense, but it didn't seem to me that it would.  I scanned the names of the other .md files in that directory and didn't see an obvious place to put something like this.  Willing to take a look and see if there's some specific place you think I should put more information?